src/server/engine_shard_set.cc

```diff
@@ -39,8 +49,260 @@ using absl::GetFlag;
 
 namespace {
 
+constexpr DbIndex DEFAULT_DB_INDEX = 0;
+
+enum class ObjectTypeForScan : uint8_t {
+  ALL = 0x1,
+  STRING = 0x2,
+  LIST = 0x4,
+  SET = 0x8,
+  ZSET = 0x10,
+  HASH = 0x20,
+  MODULE = 0x40,
+  STREAM = 0x80
+};
+
+auto operator&(uint8_t left, ObjectTypeForScan right) -> uint8_t {
+  return left & static_cast<uint8_t>(right);
+}
+
+auto operator|(uint8_t left, ObjectTypeForScan right) -> uint8_t {
+  return left | static_cast<uint8_t>(right);
+}
+
+auto operator&(ObjectTypeForScan left, uint8_t right) -> uint8_t {
+  return static_cast<uint8_t>(left) & right;
+}
+
+inline auto FromRedisType(int type) -> ObjectTypeForScan {
+  switch (type) {
+    case OBJ_STRING:
+      return ObjectTypeForScan::STRING;
+    case OBJ_LIST:
+      return ObjectTypeForScan::LIST;
+    case OBJ_SET:
+      return ObjectTypeForScan::SET;
+    case OBJ_ZSET:
+      return ObjectTypeForScan::ZSET;
+    case OBJ_HASH:
+      return ObjectTypeForScan::HASH;
+    case OBJ_MODULE:
+      return ObjectTypeForScan::MODULE;
+    case OBJ_STREAM:
+      return ObjectTypeForScan::STREAM;
+    default:
+      return ObjectTypeForScan::STRING;
+  }
+}
+
+auto OpTypeToString(std::uint8_t t) -> std::string {
+  std::string result;
+  if (t & ObjectTypeForScan::ALL) {
+    result = "all|";
+  }
+  if (t & ObjectTypeForScan::STRING) {
+    result += "string|";
+  }
+  if (t & ObjectTypeForScan::LIST) {
+    result += "list|";
+  }
+  if (t & ObjectTypeForScan::HASH) {
+    result += "hash|";
+  }
+  if (t & ObjectTypeForScan::MODULE) {
+    result += "module|";
+  }
+  if (t & ObjectTypeForScan::SET) {
+    result += "set|";
+  }
+  if (t & ObjectTypeForScan::ZSET) {
+    result += "zset|";
+  }
+  if (t & ObjectTypeForScan::STREAM) {
+    result += "stream|";
+  }
+  return result;
+}
+
+inline auto MatchType(int redis_type, std::uint8_t op_type) -> bool {
+  return op_type & ObjectTypeForScan::ALL || (FromRedisType(redis_type) & op_type) != 0;
+}
+
+struct MemoryDefragParams {
+  mi_heap_t* heap_ptr = nullptr;
+  DbSlice* slice = nullptr;
+  uint64_t cursor = 0;
+  uint8_t object_type = 0u | ObjectTypeForScan::STRING;  // use with map for type
+
+  MemoryDefragParams(mi_heap_t* hp, DbSlice* s) : heap_ptr{hp}, slice{s} {
+  }
+
+  void AssignType(ObjectTypeForScan t) {
+    object_type |= static_cast<uint8_t>(t);
+  }
+};
+
 vector<EngineShardSet::CachedStats> cached_stats;  // initialized in EngineShardSet::Init
 
+enum class MemDefragState : uint8_t {
```

Comment:
I prefer this than bool as it is more meaningful

Reply:
I do not understand what this means. there are too many redundant abstractions, IMHO. 
I do not think each task requires a dedicated API because this just adds to readability overload. 
Maybe this one is needed, but I am having a hard time understanding what's the contract here.